### PR TITLE
DOC: Update 1.14 notes

### PR DESCRIPTION
**Pull request by charris**
Also forward ports the 1.13.3 release notes and changelog. I suggest viewing the notes as a rendered file as there are a number of rearrangements that clutter up the diff.

### CODE REVIEW

File: doc/release/1.14.0-notes.rst [hunk 1]
```diff
@@ -342,6 +364,23 @@ This new default changes the float output relative to numpy 1.13. The old
 behavior can be obtained in 1.13 "legacy" printing mode, see compatibility
 notes above.
 
+''hermitiean`` option added to``np.linalg.matrix_rank``
```
Comment (eric-wieser) [inline doc/release/1.14.0-notes.rst]: Typo

File: doc/release/1.14.0-notes.rst [hunk 2]
```diff
@@ -515,15 +535,18 @@ identically-named field in the source array or to 0 if the source did not have
...
-computing dtype equality. For example with the dtypes
-`x = dtype({'names': ['A', 'B'], 'formats': ['i4', 'f4'], 'offsets': [0, 4]})`
-`y = dtype({'names': ['B', 'A'], 'formats': ['f4', 'i4'], 'offsets': [4, 0]})`
-now `x == y` will return `False`, unlike before. This makes dictionary-based
-dtype specifications like `dtype({'a': ('i4', 0), 'b': ('f4', 4)})` dangerous
-in python < 3.6 since dict key-order is not preserved in those versions.
+computing dtype equality. For example, with the dtypes ::
+
+    `x = dtype({'names': ['A', 'B'], 'formats': ['i4', 'f4'], 'offsets': [0, 4]})`
+    `y = dtype({'names': ['B', 'A'], 'formats': ['f4', 'i4'], 'offsets': [4, 0]})`
```
Comment (eric-wieser) [inline doc/release/1.14.0-notes.rst]: Backticks no longer needed?

File: doc/release/1.14.0-notes.rst [hunk 3]
```diff
@@ -580,3 +591,12 @@ The ``repr`` of a ``MaskedArray`` is now closer to the python code that would
...
 formatting changes, this can be disabled with the 1.13 legacy printing mode in
 order to help transition doctests.
+
+The ``repr`` of ``np.polynomial`` classes is more explicit
+----------------------------------------------------------
+It now shows the domain and window parameters as keyword arguments to make
+them more clear::
+
+    >>> np.polynomial.Polynomial(range(4))
+    Polynomial([ 0.,  1.,  2.,  3.], domain=[-1,  1], window=[-1,  1])
```
Comment (eric-wieser) [inline doc/release/1.14.0-notes.rst]: This needs updating with the new `repr`

**Comment by jakirkham** [comment]
Did PR ( https://github.com/numpy/numpy/pull/5580 ) get added to the changelog?

**Comment by charris** [comment]
@jakirkham The changelog is script generated before the release. It will be in the `1.14.0-changelog` when it is made.

**Comment by jakirkham** [comment]
Sure. Didn't know if it needed further mention since it changes behavior.

**Comment by charris** [comment]
@jakirkham It does probably need a release note. If you post an entry here I'll put it in, probably under `Compatibility Notes`.

File: doc/release/1.14.0-notes.rst [hunk 4]
```diff
@@ -2,83 +2,97 @@
...
-------------------------------------------------------------
-Without this argument, this falls back on a broken version of `np.frombuffer`
-that silently accepts and then encode unicode strings. If reading binary data
-is desired, ``frombuffer`` should be used directly.
+* Calling ``np.bincount`` with ``minlength=None`` is deprecated.
+  ``minlength=0`` should be used instead.
+
+* Calling ``np.fromstring`` with the default value of the ``sep`` argument is
+  deprecated.  When that argument is not provided, a broken version of
+  `np.frombuffer` is used that silently accepts unicode strings and -- after
```
Comment (rgommers) [inline doc/release/1.14.0-notes.rst]: double rather than single backticks for `np.frombuffer`

File: doc/release/1.14.0-notes.rst [hunk 5]
```diff
@@ -342,6 +364,28 @@ This new default changes the float output relative to numpy 1.13. The old
 behavior can be obtained in 1.13 "legacy" printing mode, see compatibility
 notes above.
 
+''hermitian`` option added to``np.linalg.matrix_rank``
```
Comment (rgommers) [inline doc/release/1.14.0-notes.rst]: incorrect ticks instead of backticks here

File: doc/release/1.14.0-notes.rst [hunk 6]
```diff
@@ -342,6 +364,28 @@ This new default changes the float output relative to numpy 1.13. The old
...
+
+``concatenate`` and ``stack`` gained an ``out`` argument
+--------------------------------------------------------
+A preallocated buffer of the desired dtype can now be used for the output of
+these functions.
+
+Support for PGI flang compiler on Windows
+-----------------------------------------
+The PGI flang compiler is a Fortran front end for LLVM released by NVIDIA under
+the Apache 2 license.
```
Comment (rgommers) [inline doc/release/1.14.0-notes.rst]: Would be good to add the magic incantation here, something like (did not check, just an example):  

    python setup.py config --compiler=clang --fcompiler=flang install

**Comment by charris** [comment]
@jakirkham I added an entry for #5580.

**Comment by jakirkham** [comment]
Thanks @charris.

File: doc/release/1.14.0-notes.rst [hunk 7]
```diff
@@ -206,37 +240,37 @@ enabling the new 1.13 "legacy" printing mode. This is enabled by calling
...
+* Float arrays printed in scientific notation no longer use fixed-precision,
+  and now instead show the shortest unique representation.
+* The ``str`` of floating-point scalars is no longer truncated in python2.
+* Non-finite complex scalars print like ``nanj`` instead of ``nan*j``.
+* ``MaskedArray`` arrays now separate printed elements with commas, always
+  print the dtype, and correctly wrap the elements of long arrays to multiple
+  lines. If there is more than 1 dimension, the array attributes are now
+  printed in a new "left-justified" printing style.
+* ``NaT`` values in datetime arrays are now properly aligned.
+* Arrays and scalars of ``np.void`` datatype are now print using hex notation.
```
Comment (ahaldane) [inline doc/release/1.14.0-notes.rst]: could we remove "are" here? or print -> printed.

File: doc/release/1.14.0-notes.rst [hunk 8]
```diff
@@ -248,7 +282,8 @@ C API changes
 ---------------------------------------------------------------------
 
 The ``UPDATEIFCOPY`` flag is deprecated, and replaced by a new flag
-``WRITEBACKIFCOPY``. Use of the flag requires a call to ``PyArray_ResolveWritebackIfCopy``.
+``WRITEBACKIFCOPY``. Use of the flag requires a call to
+``PyArray_ResolveWritebackIfCopy``.
 
 Code using ``UPDATEIFCOPY`` must be updated to account for changed semantics, as
 follows:. When ndarrays are created with either the deprecated ``UPDATEIFCOPY``
```
Comment (ahaldane) [inline doc/release/1.14.0-notes.rst]: remove period after :

**Comment by charris** [comment]
@mattip The `PyArray_ResolveWritebackIfCopy` function is not documented in `c-api.array.rst`, could you fix that?

**Comment by charris** [comment]
@mattip Just to be sure, the way I would like things to be for 1.14 (and later) is that nobody will notice the updateifcopy work unless they specifically intend to support PyPy. I don't want downstream projects to need concern themselves otherwise. So for this work I think deprecate is too strong a word. To go along with that, I think the documentation in `c-info.how-to-extend.rst` should have a separate section on PyPy compatiility, and the other bits should probably be unchanged, as writing to the new interface is *not* backwards compatible.

**Comment by mattip** [comment]
I am working on a documentation pull request.

Currently (on master) calling ``PyArrayObject`` constructor functions with ``NPY_ARRAY_UPDATEIFCOPY`` will emit a deprecation, advising to use ``NPY_ARRAY_WRITEBACKIFCOPY`` + ``PyArray_ResolveWritebackIfCopy``. Is there consensus that this (and the test) should be reverted for the foreseeable future?

**Comment by mattip** [comment]
> writing to the new interface is not backwards compatible

Using the ``NPY_WRITEBACKIFCOPY`` flag, without calling ``PyArray_ResolveWritebackIfCopy`` will produce correct results and raise a ``DeprecationWarning``. Using ``NPY_UPDATEIFCOPY`` will emit a ``DeprecationWarning`` when creating the array but not when deallocating it.

I am not sure I understood what you meant.

**Comment by charris** [comment]
@mattip The basic policy is that we cannot break the C-API, that is, extension modules compiled against earlier versions of Numpy should still work. Now there are a few api functions that are deprecated because of changes to the python api, mostly due to the transition to 64 bits, but those c-api functions are still there after all this time, they just call the newer functions internally. The current changes are a bit iffier, as they are only needed for PyPy compatibility (sorry ;), so I would actually prefer a whole new iterator rather than changing the behavior of the old one. I suspect that there is a lot of code that could be used by both the old and new iterators with small modifications. I am not familiar enough with the current state of things to make specific suggestions, so am just trying so set out the general philosophy. Perhaps this needs some more discussion, or maybe I just need some instruction. @eric-wieser, @ahaldane, @pv, @njsmith Thoughts?

**Comment by mattip** [comment]
>  I would actually prefer a whole new iterator ...

``nditer`` and ``NpyIter`` pull request #9998 was not merged, the UPDATEIFCOPY pull request #9639 which deprecates the UPDATEIFCOPY flag in favor of WRITEBACKIFCOPY + ``PyArray_ResolveWritebackIfCopy`` was. There is agreement that #9998 is not ready. Is there a place where I jumped the gun with documentation, or the documentation is unclear?

**Comment by charris** [comment]
> Is there a place where I jumped the gun with documentation, or the documentation is unclear?

Well,  I don't know, that's what I am trying to find out :) I Found the documentation in the release notes somewhat confusing, so what I have now (not committed) is
```
PyPy compatible alternative to ``UPDATEIFCOPY`` arrays
------------------------------------------------------
``UPDATEIFCOPY`` arrays are contiguous copies of existing arrays, possibly with
different dimensions, whose contents are copied back to the original array when
their refcount goes to zero and they are deallocated. Because PyPy does not use
refcounts, they do not function correctly with PyPy. NumPy is in the process of
eliminating their use internally and two new C-API functions,
``PyArray_SetWritebackIfCopyBase`` and ``PyArray_ResolveWritebackIfCopy``, have
been added together with a complimentary flag, ``WRITEBACKIFCOPY``. If PyPy
compatibility is not a concern, these new functions can be ignored. If you do
wish to pursue PyPy compatibility, more information on these functions may be
found at <missing link>```

**Comment by mattip** [comment]
@charris see PR #10166 for PyArray_ResolveWritebackIfCopy. If you reach the conclusion UPDATEIFCOPY should not be deprecated, this and other places should be changed, as well as C code and a deprecation test

**Comment by mattip** [comment]
Alternate wording for the release notes, pending a decision on deprecation/non-deprecation:
```
``UPDATEIFCOPY`` semantics deprecated in favor of ``WRITEBACKIFCOPY``
---------------------------------------------------------------------
``UPDATEIFCOPY`` arrays are contiguous copies of existing arrays, possibly with
different memory layout, whose contents are copied back to the original array when done.
The trigger for the copy was deallocation. To accommodate PyPy's garbage collection and
to properly separate functionality from memory management in  the ``array_dealloc``
function, this flag has been deprecated in favor of a new ``WRITEBACKIFCOPY`` and a
new API function ``PyArray_ResolveWritebackIfCopy`` added. 
``PyArray_SetUpdateIfCopyBase`` has likewise
been deprecated in favor of ``PyArray_SetWritebackIfCopyBase``. If an array is created
with copied data, either through use of flags or ``PyArray_Set...Base`` functions,
``PyArray_ResolveWritebackIfCopy`` should be called to copy the data back. An additional
function ``PyArray_DiscardWritebackIfCopy`` may be used instead to throw the copied data
away without copying it back
```

**Comment by charris** [comment]
I think the tutorial parts of the alternative belong elsewhere. Release notes are "what" and maybe some "why", but not so much "how" ;)

**Comment by mattip** [comment]
```
``UPDATEIFCOPY`` semantics deprecated in favor of ``WRITEBACKIFCOPY``
---------------------------------------------------------------------
``UPDATEIFCOPY`` arrays are contiguous copies of existing arrays, possibly with
different memory layout, whose contents are copied back to the original array when done.
The trigger for the copy was deallocation. To accommodate PyPy's garbage collection and
to properly separate functionality from memory management in  the ``array_dealloc``
function, this flag has been deprecated in favor of a new ``WRITEBACKIFCOPY`` flag.
Use of the flag requires a call to the new ``PyArray_ResolveWritebackIfCopy`` function.
The new ``PyArray_DiscardWritebackIfCopy`` may be used instead.

``PyArray_SetUpdateIfCopyBase`` has likewise
been deprecated in favor of ``PyArray_SetWritebackIfCopyBase``.

Calls to ``PyArray_XDECREF_ERR`` have been deprecated and
should be replaced by ``PyArray_DiscardWritebackIfCopy`` + ``Py_XDECREF```

``NPY_ARRAY_INOUT_ARRAY`` and
``NPY_ARRAY_INOUT_FARRAY`` flags should be replaced by
``NPY_ARRAY_INOUT_ARRAY2`` and ``NPY_ARRAY_INOUT_FARRAY2``, which behave
similarly but require use of the new ``WRITEBACKIFCOPY`` semantics.
```

**Comment by mattip** [comment]
but I should stop now, documentation is not my forte

**Pull request merged**